openshift-kube-apiserver/admission/customresourcevalidation/authentication/validate_authentication.go

```diff
@@ -121,14 +152,181 @@ func validateAuthenticationSpec(spec configv1.AuthenticationSpec) field.ErrorLis
+	if uid.Expression != "" {
+		childPath := path.Child("uid", "expression")
+		var err error
+		err, warn = validateCELExpression(ctx, cel, &authenticationcel.ClaimMappingExpression{
+			Expression: uid.Expression,
+		})
+
+		if len(warn) > 0 {
+			warn = fmt.Sprintf("validating %s: %s", childPath, warn)
+		}
+
+		if err != nil {
+			return field.ErrorList{field.Invalid(childPath, uid.Expression, err.Error())}, warn
+		}
+	}
+
+	return nil, warn
+}
+
+func validateExtraClaimMapping(ctx context.Context, path *field.Path, cel *celStore, extras ...configv1.ExtraMapping) (field.ErrorList, []string) {
+	errs := field.ErrorList{}
+	warns := []string{}
+	for i, extra := range extras {
+		err, warn := validateExtra(ctx, path.Child("extra").Index(i), cel, extra)
+		errs = append(errs, err...)
+		if len(warn) > 0 {
+			warns = append(warns, warn)
+		}
+	}
+	return errs, warns
+}
+
+func validateExtra(ctx context.Context, path *field.Path, cel *celStore, extra configv1.ExtraMapping) (field.ErrorList, string) {
+	childPath := path.Child("valueExpression")
+	err, warn := validateCELExpression(ctx, cel, &authenticationcel.ExtraMappingExpression{
+		Key:        extra.Key,
+		Expression: extra.ValueExpression,
+	})
+
+	if len(warn) > 0 {
+		warn = fmt.Sprintf("validating %s: %s", childPath, warn)
+	}
+
+	if err != nil {
+		return field.ErrorList{field.Invalid(childPath, extra.ValueExpression, err.Error())}, warn
+	}
+
+	return nil, warn
+}
+
+type compileResult struct {
+	err  error
+	warn string
+}
+
+func validateCELExpression(ctx context.Context, cel *celStore, accessor authenticationcel.ExpressionAccessor) (error, string) {
+	// if context has been canceled, don't try to compile any expressions
+	if err := ctx.Err(); err != nil {
+		return err, ""
+	}
+
+	result, err, _ := cel.compilingGroup.Do(accessor.GetExpression(), func() (interface{}, error) {
+		// if the expression is not currently being compiled, it might have already been compiled
+		if val, ok := cel.compiledStore.Get(accessor.GetExpression()); ok {
+			if val != nil {
+				res := val.(compileResult)
+				return res, nil
+			}
+
+			return nil, nil
+		}
+
+		// expression is not currently being compiled, and has not been compiled before (or has been long enough since it was last compiled that we dropped it).
+		// Let's compile it.
+		warningString := ""
+		compiled := make(chan error)
+		defer close(compiled)
+
+		go func() {
+			defer func() {
+				if r := recover(); r != nil {
+					// convert the panic into an error state for the expression
+					compiled <- fmt.Errorf("recovered from a panic while compiling expression %q: %v", accessor.GetExpression(), r)
+				}
+			}()
+
+			_, err := cel.compiler.CompileClaimsExpression(accessor)
+
+			compiled <- err
+		}()
+
+		warning := make(chan string, 1)
+		timer := time.AfterFunc(time.Second, func() {
+			defer close(warning)
+			warning <- fmt.Sprintf("cel expression %q took more than 1 second to compile", accessor.GetExpression())
```

Comment:
I should have noticed when I looked at this yesterday, but it's occuring to me now that the reason I suggested making this async was to be able to log the warning even in cases where the process dies before finishing compilation. If we didn't care about that, we could warn based on measuring the time elapsed during compilation.

Reply:
(It also uses less pipe that way: https://github.com/openshift/kubernetes/pull/2353#discussion_r2216124455)